src/dbnode/storage/bootstrap/bootstrapper/uninitialized/source.go

```diff
@@ -0,0 +1,156 @@
+// allowing us to configure the bootstrappers such that the commitlog
+// bootstrapper can precede the peers bootstrapper and still succeed bootstraps
+// for brand new namespaces without permitting unintentional data loss by
+// putting the noop-all or noop-none bootstrappers at the end of the process.
+// Behavior is best understood by reading the test cases for the test:
+// TestUnitializedSourceAvailableDataAndAvailableIndex
+type uninitializedSource struct {
+	opts Options
+}
+
+// NewUninitializedSource creates a new uninitialized source.
+func newUninitializedSource(opts Options) bootstrap.Source {
+	return &uninitializedSource{
+		opts: opts,
+	}
+}
+
+func (s *uninitializedSource) Can(strategy bootstrap.Strategy) bool {
+	switch strategy {
+	case bootstrap.BootstrapSequential:
+		return true
+	}
+
+	return false
+}
+
+func (s *uninitializedSource) AvailableData(
+	ns namespace.Metadata,
+	shardsTimeRanges result.ShardTimeRanges,
+	runOpts bootstrap.RunOptions,
+) result.ShardTimeRanges {
+	return s.availability(ns, shardsTimeRanges, runOpts)
+}
+
+func (s *uninitializedSource) AvailableIndex(
+	ns namespace.Metadata,
+	shardsTimeRanges result.ShardTimeRanges,
+	runOpts bootstrap.RunOptions,
+) result.ShardTimeRanges {
+	return s.availability(ns, shardsTimeRanges, runOpts)
+}
+
+func (s *uninitializedSource) availability(
+	ns namespace.Metadata,
+	shardsTimeRanges result.ShardTimeRanges,
+	runOpts bootstrap.RunOptions,
+) result.ShardTimeRanges {
+	var (
+		topoState                = runOpts.InitialTopologyState()
+		availableShardTimeRanges = result.ShardTimeRanges{}
+	)
+
+	for shardIDUint := range shardsTimeRanges {
+		shardID := topology.ShardID(shardIDUint)
+		hostShardStates, ok := topoState.ShardStates[shardID]
+		if !ok {
+			// This shard was not part of the topology when the bootstrapping
+			// process began.
+			continue
+		}
+
+		// The basic idea for the algorithm is that on a shard-by-shard basis we
+		// need to determine if the namespace is "new" in the sense that it has
+		// never been completely initialized (reached a state where all the hosts
+		// in the topology are "available").
+		// In order to determine this, we simply count the number of hosts in the
+		// "initializing" state. If this number is larger than zero, than the
+		// namespace is "new".
+		// The one exception to this case is when we perform topology changes and
+		// we end up with one extra node that is initializing which should be offset
+		// by the corresponding node that is leaving. I.E if numInitializing > 0
+		// BUT numLeaving >= numInitializing then it is still not a new namespace.
+		// See the TestUnitializedSourceAvailableDataAndAvailableIndex test for more details.
+		var (
+			numInitializing = 0
+			numLeaving      = 0
+		)
+		for _, hostState := range hostShardStates {
+			shardState := hostState.ShardState
+			switch shardState {
+			case shard.Initializing:
+				numInitializing++
+			case shard.Unknown:
```

Comment:
I don't think we should track Unknown either, it should never be a part of a topology.

Reply:
I'll do the same fallthrough thing

Reply:
done